### PR TITLE
A function to compute the downsampling error

### PR DESCRIPTION
**Pull request by eggrobin**
Doing it in numerics, so fairly heavy templatization.

### CODE REVIEW

File: geometry/named_quantities.hpp [hunk 1]
```diff
@@ -10,6 +11,27 @@
...
 
+// A trait to treat types that have a norm uniformly (using Abs for quantities
+// or double, and Norm for multivectors).
+template<typename T,
+         typename =
+             base::void_if_exists<decltype(quantities::Abs(std::declval<T>()))>>
+struct Normed : base::not_constructible {
+  using NormType = T;
+  static NormType Norm(T const& vector) {
+    return quantities::Abs(vector);
```
Comment (pleroy) [inline geometry/named_quantities.hpp]: Please move the method implementations to the body file.

File: geometry/named_quantities.hpp [hunk 2]
```diff
@@ -10,6 +11,27 @@
...
+  }
+};
+
+template<typename T>
+struct Normed<T, base::void_if_exists<decltype(std::declval<T>().Norm())>>
+    : base::not_constructible {
+  using NormType = decltype(std::declval<T>().Norm());
+  static NormType Norm(T const& vector) {
+    return vector.Norm();
```
Comment (pleroy) [inline geometry/named_quantities.hpp]: Same.

File: numerics/hermite3.hpp [hunk 3]
```diff
@@ -33,6 +36,17 @@ class Hermite3 final {
   // The result is sorted.
   BoundedArray<Argument, 2> FindExtrema() const;
 
+  // |samples| must be a container; |get_argument| and |get_value| on the
+  // elements of |samples| must return |Argument| and |Value| respectively
+  // (possibly by reference or const-reference)
+  // Returns the largest error (in the given |norm|) between this polynomial and
+  // the given |samples|.
+  template<typename Samples, typename GetArgument, typename GetValue>
+  typename Normed<Difference<Value>>::NormType LInfinityError(
```
Comment (pleroy) [inline numerics/hermite3.hpp]: Would the infinity symbol work in an identifier?

File: numerics/hermite3_body.hpp [hunk 4]
```diff
@@ -63,6 +64,21 @@ BoundedArray<Argument, 2> Hermite3<Argument, Value>::FindExtrema() const {
       arguments_.first, a1_, 2.0 * a2_, 3.0 * a3_);
 }
 
+template<typename Argument, typename Value>
+template<typename Samples, typename GetArgument, typename GetValue>
+typename Normed<Difference<Value>>::NormType
+Hermite3<Argument, Value>::LInfinityError(Samples const& samples,
+                                          GetArgument get_argument,
+                                          GetValue get_value) const {
```
Comment (pleroy) [inline numerics/hermite3_body.hpp]: Pass the functors by reference, you are not moving them.

File: numerics/hermite3_test.cpp [hunk 5]
```diff
@@ -82,5 +95,50 @@ TEST_F(Hermite3Test, Conditioning) {
             h.EvaluateDerivative(t0_ + 19418861.806896236 * Second));
 }
 
+TEST_F(Hermite3Test, OneDimensionalInterpolationError) {
+  std::vector<std::pair<double, double>> samples;
+  for (double i = 2; i < 10; i += 1) {
```
Comment (pleroy) [inline numerics/hermite3_test.cpp]: Iterating on double :-O

File: numerics/hermite3_test.cpp [hunk 6]
```diff
@@ -82,5 +95,50 @@ TEST_F(Hermite3Test, Conditioning) {
...
+      Eq(1 / 16.0));
+}
+
+TEST_F(Hermite3Test, ThreeDimensionalInterpolationError) {
+  std::vector<std::pair<Instant, Position<World>>> samples;
+  Instant const tmax = t0_ + π / 2 * Second;
+  AngularFrequency const ω = 1 * Radian / Second;
+  for (Instant t = t0_; t <= tmax; t += 1 / 32.0 * Second) {
+    samples.push_back(
+    {t, World::origin + Displacement<World>({Cos(ω * (t - t0_)) * Metre,
```
Comment (pleroy) [inline numerics/hermite3_test.cpp]: Indent += 4

File: numerics/hermite3_test.cpp [hunk 7]
```diff
@@ -82,5 +95,50 @@ TEST_F(Hermite3Test, Conditioning) {
...
+       World::origin + Displacement<World>({0 * Metre, 1 * Metre, 0 * Metre})},
+      /*derivatives=*/
+      {Velocity<World>(
+           {0 * Metre / Second, 1 * Metre / Second, 0 * Metre / Second}),
+       Velocity<World>(
+           {-1 * Metre / Second, 0 * Metre / Second, 0 * Metre / Second})});
+  EXPECT_THAT(
+      not_a_quartic.LInfinityError(
+          samples,
+        /*get_argument=*/[](auto&& pair) -> auto&& { return pair.first; },
```
Comment (pleroy) [inline numerics/hermite3_test.cpp]: Bad indent (2x).

File: numerics/hermite3_test.cpp [hunk 8]
```diff
@@ -82,5 +95,50 @@ TEST_F(Hermite3Test, Conditioning) {
...
+        /*get_value=*/[](auto&& pair) -> auto&& { return pair.second; }),
+      AllOf(Gt(1 * Centi(Metre)), Lt(2 * Centi(Metre))));
+}
+
```
Comment (pleroy) [inline numerics/hermite3_test.cpp]: Should we have a test with a quartic?

File: numerics/hermite3.hpp (hunk 3 above)
Comment (eggrobin) [reply numerics/hermite3.hpp]: No, we had the same issue with InfinitePast and InfiniteFuture.

File: geometry/named_quantities.hpp (hunk 1 above)
Comment (eggrobin) [reply geometry/named_quantities.hpp]: I don't know to write the definitions out of line with this fancy specialization.

File: numerics/hermite3_body.hpp (hunk 4 above)
Comment (eggrobin) [reply numerics/hermite3_body.hpp]: Done.

File: numerics/hermite3_test.cpp (hunk 5 above)
Comment (eggrobin) [reply numerics/hermite3_test.cpp]: ¯\_(ツ)_/¯

File: numerics/hermite3_test.cpp (hunk 6 above)
Comment (eggrobin) [reply numerics/hermite3_test.cpp]: Done.

File: numerics/hermite3_test.cpp (hunk 7 above)
Comment (eggrobin) [reply numerics/hermite3_test.cpp]: Done.

File: numerics/hermite3_test.cpp (hunk 8 above)
Comment (eggrobin) [reply numerics/hermite3_test.cpp]: We do (it's the one above), but perhaps I shouldn't call the function that tries to match a circle `not_a_quartic`... :-p

File: geometry/named_quantities.hpp (hunk 1 above)
Comment (pleroy) [reply geometry/named_quantities.hpp]: See https://godbolt.org/g/GXy9jk.

Comment (eggrobin) [reply geometry/named_quantities.hpp]: Done.

File: geometry/named_quantities.hpp (hunk 2 above)
Comment (eggrobin) [reply geometry/named_quantities.hpp]: Done.

**Comment by eggrobin** [comment]
This is part of the work on #228.

**Pull request merged**